[Review]feat(pubsub): KeyStorage and SetSecurityKeys

Muddasir-Shakil opened the pull request:
This PR adds the basic support for SKS feature. It adds a Keystorage to store and maintain the life cycle of the keys used to secure the message and implementation of setSecurityKeys method node to push the keys to Publishers and Subscribers.

Thanks to the author of #3255, I based my work on this PR.

The ua_pubsub_keystorage provides the keystorage structure and interface to interact with the keystorage. The setSecurityKeys
method callback is added to ua_pubsub_ns0. The unit tests are added to tests/pubsub/check_pubsub_keystorage.c and tests/pubsub/check_pubsub_sks_push.c

## What is supported?
- [x] Keystorage to store the keyLists (current and future keys)
- [x] Storing/updating keys in the keyStorage
- [x] Automatic rollover to next key from local keysList
- [x] Setting keys for encryption/decryption and singing/verification in the channel contexts of the associated PubSub Groups 
- [x] SetSecurityKeys method node callback to push the keys
- [x] unit tests for keystorage and setSecurityKey Method Node
- [x] Integration of sks build and keystorage in the CI pipeline
## What is missing ?

- Checking user credentials against SecurityGroup user access rights when calling setSecurityKeys
- AddSecurityGroup Method Node 
- RemoveSecurityGroup Method Node
- GetSecurityGroup Method Node
- GetSecurityKeys Method Node
- Write SecurityGroupId and SecurityKeyServices Url in PubSub Group Object Node

andreasebner commented on .github/workflows/build_linux.yml:
```diff
@@ -41,6 +41,9 @@ jobs:
           - build_name: "PubSub Encryption (MbedTLS) Build & Unit Tests (gcc)"
             cmd_deps: sudo apt-get install -y -qq libmbedtls-dev
             cmd_action: unit_tests_encryption_mbedtls_pubsub
+          - build_name: "PubSub SKS Build & Unit Tests (gcc)"
+            cmd_deps: sudo apt-get install -y -qq libmbedtls-dev
+            cmd_action: unit_tests_pubsub_sks
```
First of all, great work! Is there a special reason to have a dedicated github runner? I would preffer to integrate these tests into one of the exisitng encryption tests.

Muddasir-Shakil replied:
Thank you for the review. With SKS support we will have two mechanims to set the security keys. If Encryption is ON then we have no problem but if Encryption and SKS is ON then the keys has to be set from the keystorage. Also in Pull model, which will be the next PR, it requires the keys to be fetched from SKS before writer/reader group can be set operational. When they are operational, check the current key in key storage and set the encrypting keys.

andreasebner commented on src/pubsub/ua_pubsub_keystorage.c:
```diff
@@ -0,0 +1,465 @@
+static void
+UA_PubSubKeyStorage_clearKeyList(UA_PubSubKeyStorage *keyStorage) {
+    if(keyStorage->keyList) {
+
+        UA_PubSubKeyListItem *iterator = keyStorage->firstItem;
+        UA_PubSubKeyListItem *head = iterator;
+        while(iterator != NULL) {
+            if(iterator->key.data) {
+                UA_free(iterator->key.data);
+                head = iterator;
+                iterator = iterator->next;
+                UA_free(head);
+            } else
+                break;
+        }
+        keyStorage->keyListSize = 0;
+
```
Please avoid this empty lines (check it for the whole PR). Sometimes the seperation make sense to make the code more readable, but this should be the exception.

Muddasir-Shakil replied:
Thank you for pointing it out. I will check the empty lines, but in most cases I add to keep the context of the code clear.

andreasebner commented on src/server/ua_server_internal.h:
```diff
@@ -172,6 +173,10 @@ struct UA_Server {
     UA_PubSubManager pubSubManager;
 #endif
 
+#if defined(UA_ENABLE_PUBSUB_ENCRYPTION) && defined(UA_ENABLE_PUBSUB_SKS)
+    LIST_HEAD(PubSubKeyList, UA_PubSubKeyStorage) pubSubKeyList;
+#endif
+
```
Could you please move the "pubSubKeyList" into the "pubSubManager"?

Muddasir-Shakil replied:
Yes, I also realized it later that it is better suited in PubSubManager. I did the same thing for SecurityGroups  managed by the PubSub Manager ( not yet Pushed)

andreasebner reviewed (commented):
Creat work! Added some minor comments. Could you also please make an interactive rebase to sync your version with the latest master and reduce the number of commits?

Muddasir-Shakil commented:
> Great work! Added some minor comments. 

Thank you @andreasebner  for the comments

>Could you also please make an interactive rebase to sync your version with the latest master and 

Yes, I will do. While I was working on SecurityGroups and SKS Pull, I also found some minor improvements that I will try to merge here.
>reduce the number of commits?

Now that is reviewed I can squash relatable parts like all the unit tests into one commit for a related feature. This should shorten the commit numbers

Muddasir-Shakil commented:
Hi @andreasebner 
I update the PR according to your comments and merge the commits to shorten the number of commits. Some additional minor improvements are also done which involve. 

1. importing key to channel context of the select PubSub Group or all the PubSubGroups with matching SecurityGroupId
2. fixed the initialize key storage method to set the maximum key size from maximum future and past key size.
3. fixed memory leak issues when adding an existing keyStorage to the PubSubGroup.
4. switched to fakesleep for UA_Sleep_ms in unit tests.
5. Synced with upstream master

jpfr commented on src/pubsub/ua_pubsub_keystorage.h:
```diff
@@ -0,0 +1,403 @@
+ * @return UA_StatusCode return status code
+ */
+UA_StatusCode
+UA_PubSubKeyStorage_initialize(UA_Server *server, const UA_String *securityGroupId,
```
Shorten to _init to stay within the nomenclature of the SDK.

Muddasir-Shakil replied:
done

jpfr commented on src/pubsub/ua_pubsub_keystorage.h:
```diff
@@ -0,0 +1,403 @@
+    /**
+     * the last key item in the storage
+     */
+    UA_PubSubKeyListItem *lastItem;
```
If you need a list with first and last element known, use the TAILQ structure from /deps/open61541_queue.h.

jpfr commented on src/pubsub/ua_pubsub_keystorage.c:
```diff
@@ -0,0 +1,530 @@
+        UA_PubSubKeyListItem *head = iterator;
+        while(iterator != NULL) {
+            if(iterator->key.data) {
+                UA_free(iterator->key.data);
```
UA_ByteString_clear.
That covers the difference between "empty string" and "null string" in the SDK.

jpfr commented on src/pubsub/ua_pubsub_keystorage.c:
```diff
@@ -0,0 +1,530 @@
+                iterator = iterator->next;
+                UA_free(head);
+            } else
+                break;
```
Is there a special semantics to a listitem with empty key?
Add a comment to explain why you break.
Or don't break...

Muddasir-Shakil replied:
Not needed anymore after changing to TAILQ list

jpfr commented on src/pubsub/ua_pubsub_keystorage.c:
```diff
@@ -0,0 +1,530 @@
+}
+
+void
+UA_PubSubKeyStorage_deleteMembers(UA_Server *server, UA_PubSubKeyStorage *keyStorage) {
```
Name this _clear.

Muddasir-Shakil replied:
done

jpfr commented on src/pubsub/ua_pubsub_keystorage.c:
```diff
@@ -0,0 +1,530 @@
+        return;
+
+    /*Remove callback*/
+    if(server) {
```
You can assume this is only called with the server.
It is internal API where we can prevent misuse.
You can add a UA_assert(server) toe help the analysis in the CI.

Muddasir-Shakil replied:
added UA_Assert(server) at the start of function

jpfr commented on src/pubsub/ua_pubsub_keystorage.c:
```diff
@@ -0,0 +1,530 @@
+    if(!keyStorage)
+        return;
+
+    UA_PubSubKeyStorage_deleteMembers(server, keyStorage);
```
If this is the only place that uses UA_PubSubKeyStorage_deleteMembers, then the indirection can be removed.
Just define UA_PubSubKeyStorage_delete in that case.

Muddasir-Shakil replied:
deleteMembers is now changed to clear and delete is kept because clear does not free the keystorage pointer.

jpfr commented on src/pubsub/ua_pubsub_keystorage.c:
```diff
@@ -0,0 +1,530 @@
+}
+
+UA_StatusCode
+UA_PubSubKeyStorage_getKeyByKeyID(const UA_UInt32 keyId, UA_PubSubKeyListItem *firstItem,
```
In the sense of a "quasi object-oriented" API, forward a UA_PubSubKeyStorage pointer to this method.
Instead of forwarding only elements from a UA_PubSubKeyStorage.
That also allows us the change the internal definitions of UA_PubSubKeyStorage without changing this API.

This applies to all methods with the UA_PubSubKeyStorage_ prefix.

Muddasir-Shakil replied:
Whereever possible changed to passing PubSubKeyStorage pointer.

jpfr reviewed (commented):
This is already veyr high quality.
Some comments inline.

Happy to see this merged soon.

Muddasir-Shakil commented on src/pubsub/ua_pubsub_keystorage.c:
```diff
@@ -0,0 +1,530 @@
+            break;
+        }
+    }
+    return retval;
```
Thanks for the comments.
In this case, it returns the bad statuscode. Now I added a unit test to verify that and changed it to BADNOTFOUND.

Muddasir-Shakil commented on src/pubsub/ua_pubsub_keystorage.h:
```diff
@@ -0,0 +1,403 @@
+ * @param keyStorage keyStorage to delete
+ */
+void
+UA_PubSubKeyStorage_deleteMembers(UA_Server *server, UA_PubSubKeyStorage *keyStorage);
```
done

Muddasir-Shakil commented:
Hello @jpfr,  
I updated the PR according to your comments. The major change involve is switching to TAILQ for keyList in the keyStorage. Also whereever possible, I change the internal API to forward PubSubkeyStorage pointer. Meanwhile, I have readied the Patch queue for the securityGroups, sks pull and sks client. I will rebase and create PR once this PR gets into upstream. 

Looking forward to hear from you.

jpfr commented on src/pubsub/ua_pubsub_keystorage.h:
```diff
@@ -0,0 +1,396 @@
+ * @return UA_PubSubKeyStorage*
+ */
+UA_PubSubKeyStorage *
+UA_PubSubKeyStorage_new(void);
```
A separate _new is unnecessary complexity.
All it does is malloc and zero out the memory.
You can just use UA_calloc in the few places that create a new storage.

Muddasir-Shakil replied:
removed and replaced with UA_calloc everywhere

jpfr commented on src/pubsub/ua_pubsub_ns0.c:
```diff
@@ -1633,6 +1634,101 @@ removeDataSetWriterAction(UA_Server *server,
+        return UA_STATUSCODE_BADTOOMANYARGUMENTS;
+
+    /*check for types*/
+    if(input[0].type != &UA_TYPES[UA_TYPES_STRING] ||     /*SecurityGroupId*/
```
Also check whether the variants contain a scalar.
If they contain an empty array (of that type), then oob memory is accessed.
We have `UA_Variant_isScalarType` that does everything in one go.

Muddasir-Shakil replied:
changed to hasScalarType

jpfr commented on src/pubsub/ua_pubsub_ns0.c:
```diff
@@ -1633,6 +1634,101 @@ removeDataSetWriterAction(UA_Server *server,
+    if(!server || !input)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    /* TODO: Check user credentials*/
```
This comment can leave people confused whether this is inherently insecure as an implementation.
Rather write that the user credentials are checked in the AccessControl plugin *before* this callback is executed.
It is up to the user to configure his AccessControl plugin in a way that suits the use case.

Muddasir-Shakil replied:
I added a @note before this callback and removed the comment.

jpfr commented on src/pubsub/ua_pubsub_keystorage.h:
```diff
@@ -0,0 +1,396 @@
+ * @param groupId of the PubSubGroup
+ */
+void
+UA_PubSubKeyStorage_removeKeyStorage(UA_Server *server, UA_String securityGroupId);
```
In nearly all cases where this is used, you first do a lookup, then check if `ks != NULL`, then call removeKeyStorage.
Just forward the ks pointer to `removeKeyStorage` instead of doing a duplicate lookup.

jpfr commented on src/pubsub/ua_pubsub_keystorage.h:
```diff
@@ -0,0 +1,396 @@
+ * @param keyStorage the pointer to the keystorage
+ */
+void
+UA_moveToNextkeyCallback(UA_Server *server, UA_PubSubKeyStorage *keyStorage);
```
Why no `PubSubKeyStorage` in the prefix?

Muddasir-Shakil replied:
added the PubSubKeyStorage prefix

jpfr commented on src/pubsub/ua_pubsub_keystorage.h:
```diff
@@ -0,0 +1,396 @@
+ * @return UA_StatusCode the return status
+ */
+UA_StatusCode
+UA_PubSubKeyStorage_addMoveToNextKeyCallback(UA_Server *server,
```
It is imho clearer to write `Rollover` instead of `MoveToNextKey`.

Muddasir-Shakil replied:
replaced with _keyRolloverCallback()

jpfr commented on src/pubsub/ua_pubsub_keystorage.h:
```diff
@@ -0,0 +1,396 @@
+ * @return UA_StatusCode return status code
+ */
+UA_StatusCode
+UA_PubSubKeyStorage_getPubSubSecurityPolicyByUri(UA_Server *server,
```
`UA_Server_findPubSubSecurityPolicy`

jpfr replied:
This can then be reused in other places as well.

Muddasir-Shakil replied:
should it also be public? UA_EXPORT?

Muddasir-Shakil replied:
Changed it to UA_Server_findPubSubSecurityPolicy

jpfr commented on src/pubsub/ua_pubsub_keystorage.h:
```diff
@@ -0,0 +1,396 @@
+ * @return Pointer to the keystorage on success, null pointer on failure
+ */
+UA_PubSubKeyStorage *
+UA_PubSubKeyStorage_getKeyStorageWithSecurityGroupId(UA_Server *server,
```
`UA_Server_findKeyStorage`

Muddasir-Shakil replied:
should it also be public? UA_EXPORT?

jpfr replied:
Since the key storage structure is not public, no.

Muddasir-Shakil replied:
Changed it to UA_Server_findKeyStorage

jpfr commented on src/pubsub/ua_pubsub_manager.c:
```diff
@@ -773,6 +777,14 @@ UA_PubSubManager_delete(UA_Server *server, UA_PubSubManager *pubSubManager) {
     }
+
+#if defined(UA_ENABLE_PUBSUB_ENCRYPTION) && defined(UA_ENABLE_PUBSUB_SKS)
+    UA_PubSubKeyStorage *keyStorage, *keyStorageTemp;
```
If you call the variables `ks, ksTmp`, then the foreach fits into a single line.

Muddasir-Shakil replied:
shorten the pointer names

jpfr commented on src/pubsub/ua_pubsub_keystorage.h:
```diff
@@ -0,0 +1,396 @@
+                                          UA_UInt64 *callbackID);
+
+/**
+ * @brief It takes the current Key data, divide it into singing key, encrypting key and
```
singing key

jpfr commented on src/pubsub/ua_pubsub_keystorage.h:
```diff
@@ -0,0 +1,396 @@
+ * @return UA_StatusCode return status code
+ */
+UA_StatusCode
+UA_PubSubKeyStorage_importKeyToChannelContext(UA_Server *server, const UA_NodeId pubSubGroupId,
```
This is used in only one place and does not have to be exposed.
It makes sense that the keys are loaded when the rollover happens (and only then).
Make the function `static`, i.e. defined only in the compilation unit.
Also, the only place that uses it calls with NODEID_NULL.

Muddasir-Shakil replied:
There are technically three scenarios for this method here

1. When the SetSecurityKeys Method is called, it passes the current and future key. The current key is immediately imported into channel context with this method.
2. When keylifetime expires, the Rollover moves the currentItem pointer to the next key and then imports that key to the channelcontext. The first iteration happens after the first keylifetime period expires, which can take any amount of time.
3. When keys are fetched using Pull Model (GeSecurityKeys Method) and stored in the keyStorage and the currentItem points to the current key. Then Writer/Reader groups can be into operational at random times. Then the current key is imported from KeyStorage. It is also possible that the Rollover has already done multiple iterations before setting into Operational state, in this Channel Context will be update with same key material.

These are the reasons I set it non static.

NODEID_NULL, is used to set the keys for all the related Writer/Reader Groups. However, when a nodeId is given only that Reader/Writer Group is update. It is used in the next patches related to SKS client.

jpfr replied:
Ok. So there is additional usage that will only come with the future PR.
I think the name _activateKey makes more sense.
Import sounds like importing from a file.

Muddasir-Shakil replied:
changed it to _activateKeyToChannelContext()

Muddasir-Shakil commented on CMakeLists.txt:
```diff
@@ -388,6 +388,9 @@ mark_as_advanced(UA_ENABLE_PUBSUB_MONITORING)
 option(UA_ENABLE_PUBSUB_BUFMALLOC "Enable allocation with static memory buffer for time critical PubSub parts" OFF)
 mark_as_advanced(UA_ENABLE_PUBSUB_BUFMALLOC)
 
+option(UA_ENABLE_PUBSUB_SKS "Enable Security Key Service support for publisher and subscriber" OFF)
```
should we also mark it Experimental until all the patches are merged related to SKS?

jpfr replied:
I'd go so far to add a warning message in cmake.
That the feature is currently merged but not yet ready.

Muddasir-Shakil replied:
Added the Check for UA_PUBSUB_ENCRYPTION and warning text.

jpfr commented on src/pubsub/ua_pubsub_keystorage.c:
```diff
@@ -0,0 +1,498 @@
+
+#if defined(UA_ENABLE_PUBSUB_ENCRYPTION) && defined(UA_ENABLE_PUBSUB_SKS) /* conditional compilation */
+
+#define UA_CURRENTKEY_COUNT 1
```
Yeah. A variable in the local scope is preferrable.
Global defines are a good idea in general.
But in that case it is just a headache to do the scrolling, just to find that the current count is always "1".

Muddasir-Shakil replied:
Removed from global scope to the local method scope

Muddasir-Shakil commented on src/pubsub/ua_pubsub_keystorage.h:
```diff
@@ -0,0 +1,396 @@
+ * @param keyStorage keyStorage to delete
+ */
+void
+UA_PubSubKeyStorage_clear(UA_Server *server, UA_PubSubKeyStorage *keyStorage);
```
removed the indirection.

jpfr commented:
Nice! Merged.
Looking forward to the next PR.
